Explicitly set a host when using NodePort Service

andersbogsnes opened the pull request:
Previously, if the service type was a NodePort, KubeCluster would attempt to discover the IP address by listing nodes. This is a set of permissions that not all users will have, so this PR adds the ability to explicitly set a known host.

Closes #348

andersbogsnes commented:
I'm unsure about how to test this implementation, as I can't see a means of getting the node IP address when using pytest-kind. 

Even `docker inspect` gives me a blank IP Adress

jacobtomlinson commented on .gitignore:
```diff
@@ -138,3 +138,6 @@ credentials.csv
 *.pub
 *.rdp
 *_rsa
+
+# IDE settings
+.idea
```
I usually recommend folks put this kind of thing in a [global gitignore file](https://stackoverflow.com/questions/7335420/global-git-ignore).

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -578,6 +578,7 @@ async def _start(self):
                     "idle_timeout": self._idle_timeout,
                     "service_wait_timeout_s": self._scheduler_service_wait_timeout,
                     "pod_template": self.scheduler_pod_template,
+                    "host": self.host,
```
The `host` kwarg and attribute is for setting the host IP when running `deploy_mode="local"`. I think we need to add a new kwarg called something like `nodeport_host`. Perhaps we also need to update the docstring to be clear that `host` is for the local deploy mode only.

jacobtomlinson reviewed (commented):
Thanks so much for raising this @andersbogsnes.

I think for testing it will be tricky as the Kubernetes cluster is inside a docker container and I expect it will not be possible to access the nodeport directly. I would perhaps add some check that the `external_address` attribute gets set correctly when the nodeport IP is provided. But I do not expect to actually be able to connect to it.

I've added some comments below because you seem to be reusing the `host` kwarg which is used for a different purpose.

andersbogsnes commented:
Thanks for the review! I'll focus on testing some verifications that the correct values have been set.

In terms of the API - since we don't want to reuse the host arg, and we can currently only set scheduler type in the config, I see two options.

- Set scheduler-type and NodePort-host in the config
- Have both available as kwargs in the __init__. 

Personal preference is #2, but I fully appreciate the danger of a million-param __init__

jacobtomlinson commented:
The `host` arg is used for a different purpose. I see not harm in adding an additional kwarg for this. We should also support this in the config.

andersbogsnes commented:
I've implemented the two keywords nodeport_host and scheduler_service_type. I've modified the test to simulate calling KubeCluster._start() as closely as possible, as actually calling it will cause the test to fail, since it can't connect to the Kind cluster with a NodePort

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -315,7 +318,11 @@ class KubeCluster(SpecCluster):
-        Listen address for local scheduler.  Defaults to 0.0.0.0
+        Listen address for local scheduler. Only used when deploy_mode == "local". Defaults to
+        0.0.0.0
+    nodeport_host: str
+        When scheduler_service_type == ``"NodePort"``, override KubeCluster's default behaviour
+        of listing nodes by specifying a nodeport_host.
```
May I suggest a more user focused docstring here. 

```suggestion
    nodeport_host: str
        When scheduler_service_type == ``"NodePort"``, select a specific node IP to connect to.
        When this is not set the node IP will be autodetected.
```

jacobtomlinson commented on dask_kubernetes/core.py:
```diff
@@ -415,6 +426,8 @@ def __init__(
         security=None,
         scheduler_service_wait_timeout=None,
         scheduler_pod_template=None,
+        scheduler_service_type="ClusterIP",
```
This should be set to `None` as the default is already set in the configuration YAML.

jacobtomlinson commented on dask_kubernetes/utils.py:
```diff
@@ -53,8 +53,10 @@ async def get_external_address_for_scheduler_service(
     elif service.spec.type == "NodePort":
-        nodes = await core_api.list_node()
-        host = nodes.items[0].status.addresses[0].address
+        host = nodeport_host
+        if host is None:
+            nodes = await core_api.list_node()
+            host = nodes.items[0].status.addresses[0].address
```
Small nit to improve readibility.

```suggestion
        if nodeport_host is not None:
            host = nodeport_host
        else:
            nodes = await core_api.list_node()
            host = nodes.items[0].status.addresses[0].address
```

jacobtomlinson commented on dask_kubernetes/tests/test_async.py:
```diff
@@ -390,6 +392,61 @@ async def test_constructor_parameters(k8s_cluster, pod_spec):
+        cluster.pod_template, pod_type="scheduler"
+    )
+    cluster.scheduler_pod_template.spec.containers[0].args = ["dask-scheduler"]
+    cluster.scheduler_pod_template = clean_pod_template(
+        pod_template=cluster.scheduler_pod_template, pod_type="scheduler"
+    )
+    await ClusterAuth.load_first(cluster.auth)
+    cluster._generate_name = escape(
+        cluster._generate_name.format(
+            user=getpass.getuser(), uuid=str(uuid.uuid4())[:10]
+        )
+    )
+    cluster.scheduler_pod_template = cluster._fill_pod_templates(
+        cluster.scheduler_pod_template, pod_type="scheduler"
+    )
+
+    scheduler = Scheduler(
+        idle_timeout=cluster._idle_timeout,
+        service_wait_timeout_s=cluster._scheduler_service_wait_timeout,
+        pod_template=cluster.scheduler_pod_template,
+        nodeport_host=cluster.nodeport_host,
+        scheduler_service_type=cluster._scheduler_service_type,
+        cluster=cluster,
+        core_api=kubernetes.client.CoreV1Api(),
+        policy_api=kubernetes.client.PolicyV1beta1Api(),
+        namespace=cluster.namespace,
+        loop=cluster.loop,
+    )
+
+    scheduler.service = await scheduler._create_service(
+        scheduler.kwargs["scheduler_service_type"]
+    )
+
+    external_address = await get_external_address_for_scheduler_service(
+        scheduler.core_api,
+        scheduler.service,
+        nodeport_host=scheduler.kwargs["nodeport_host"],
+    )
+    assert external_address == "tcp://10.10.10.10:8786"
+    assert scheduler.service.spec.type == "NodePort"
```
I appreciate how in depth you've gone here with this test but given that this had gone so deep into the guts of the library I worry that it will be brittle and break if we make internal changes.

Perhaps it would be best to focus on testing the `get_external_address_for_scheduler_service` in isolation instead.

```suggestion
    from dask_kubernetes.objects import make_service_from_dict
    
    service = make_service_from_dict({"spec": {"ports": [{"port": 8786, "name": "comm"}], "type": "NodePort"}})
    
    external_address = await get_external_address_for_scheduler_service(
        None, service, nodeport_host="10.10.10.10",
    )
    assert external_address == "tcp://10.10.10.10:8786"
```

andersbogsnes replied:
Thanks! Couldn't figure out how to create an instance of the service without going through the cluster - `make_service_from_dict` was the missing piece :smile:

jacobtomlinson commented:
Hey @andersbogsnes. This PR is still marked as a draft. Is it ready for a full review yet?

andersbogsnes commented:
@jacobtomlinson Thanks for the ping - had left this one a bit fallow! I believe so, incorporated your previous suggestions, though could never get through the CI tests - they would hang and time out, despite working fine locally

jacobtomlinson commented:
Looks like a valid failure in CI now related to this PR. @andersbogsnes would you mind taking a look?

jacobtomlinson commented:
Sorry we never got to merging this. The project has changed a lot since this was raised so going to close it out.
